azext_iot/common/utility.py

```diff
@@ -351,6 +355,36 @@ def unpack_msrest_error(e):
     return op_err
 
 
+def handle_service_exception(e):
+    from azure.cli.core.azclierror import (
+        AzureInternalError,
+        AzureResponseError,
+        BadRequestError,
+        ForbiddenError,
+        ResourceNotFoundError,
+        UnauthorizedError,
+    )
+    op_status = None
+    err = unpack_msrest_error(e)
+    try:
+        op_status = getattr(e.response, 'status_code', None)
+        if op_status in [400, 409]:
+            raise BadRequestError(err)
+        if op_status == 401:
+            raise UnauthorizedError(err)
+        if op_status == 403:
+            raise ForbiddenError(err)
+        if op_status == 404:
+            raise ResourceNotFoundError(err)
+        if op_status in [500, 503, 504]:
```

Comment:
Should this list include all 5xx values?

Reply:
The intent was to just include 5XX error codes in this group that were actual internal service errors (500/503/504*) as `AzureInternalError` - and leave the rest to generic `AzureResponseErrors` which is more of a fallback.

\* 504 / 502 is iffy (gateway timeout / bad gateway) because they don't necessarily directly indicate errors in the service, but it could be an error passed from the resource itself when the service endpoint is acting as a gateway.

I can add 501 (not implemented), 502 (bad gateway), and 505 (http version not supported) here to be more inclusive.

@digimaun  - any thoughts?

Reply:
To summarize an offline discussion, based on the documentation in the [ `azclierror` class](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli-core/azure/cli/core/azclierror.py), we're going to have all 500-599 errors raise an `AzureInternalError`, and we're also moving 409 errors from `BadRequestError` to `AzureResponseError`, as the documentation for `BadRequestError` explicitly states only 400 responses.